test/jdk/java/io/Serializable/ClassSerialization/ArchivedClassesTest.java

```diff
@@ -0,0 +1,503 @@
+                        .append('\n');
+            }
+            out.append(clazz.getName()).append(": ");
+            HexFormat.of().formatHex(out, serializedClass(clazz));
+            out.append("\n");
+        }
+    }
+
+    // Return byte array of a serialized Class object
+    private static byte[] serializedClass(Class<?> clazz) {
+        try {
+            var byteStream = new ByteArrayOutputStream();
+            try (var objStream = new ObjectOutputStream(byteStream)) {
+                objStream.writeObject(clazz);
+            }
+            return byteStream.toByteArray();
+        } catch (IOException ioe) {
+            throw new UncheckedIOException(ioe);
+        }
+    }
+
+    /**
+     * Scans all classes in the JDK for those matching a certain set of criteria.
+     * Scanning uses ModuleFinder.ofSystem to locate classes.
+     * Classes are matched using the following criteria:
+     * <p>
+     *  - is a public or protected class
+     *  - not module-info
+     *  - resides in a module that's open to or exported to the unnamed module
+     * <p>
+     * `findAll` returns a stream of classes
+     */
+    static class ClassScanner {
+        private static final ClassLoader LOADER = ClassScanner.class.getClassLoader();
+        private static final Module UNNAMED = LOADER.getUnnamedModule();
+
+        private static Optional<Class<?>> findClass(String name) {
+            try {
+                Class<?> clazz = Class.forName(name, false, LOADER);
+                return Optional.of(clazz);
+            } catch (ClassNotFoundException | ExceptionInInitializerError |
+                     NoClassDefFoundError | IllegalAccessError ex) {
+                return Optional.empty();
+            }
+        }
+
+        private static boolean isPublicOrProtected(Class<?> clazz) {
+            return (clazz.getModifiers() & (Modifier.PUBLIC | Modifier.PROTECTED)) != 0;
+        }
+
+        private static boolean isInExportedOrOpenPkg(Class<?> clazz) {
+            String pkg = clazz.getPackageName();
+            Module mod = clazz.getModule();
+            return mod.isExported(pkg, UNNAMED) || mod.isOpen(pkg, UNNAMED);
+        }
+
+        private static Stream<String> getClassFiles(ModuleReader f) {
+            try {
+                return f.list();
+            } catch (IOException ioe2) {
+                throw new UncheckedIOException(ioe2);
+            }
+        }
+
+        private static ModuleReader getModuleReader(ModuleReference f) {
+            try {
+                return f.open();
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        }
+
+        /**
+         * Scans classes from ModuleFinder.ofSystem().find("java.se") and returns all classes.
+         * Classes are public or protected in an exported or open package.
+         * @return Stream of matching classes
+         */
+        public static Stream<Class<?>> findAll() {
```

Comment:
Does this test have to be limited to the standard modules? I would think it should find all classes in packages that are exported unconditionally (to all modules).